### PR TITLE
#439 Freecalls - State service checks for free calls and also Authentication checks on Signature

### PR DESCRIPTION
**Pull request by anandrgitnirman**
#439 

Make state service calls to determine how many free calls are left , 
and also Authentication checks on Signature

### CODE REVIEW

File: authutils/auth_service.go [hunk 1]
```diff
@@ -83,6 +85,19 @@ func CompareWithLatestBlockNumber(blockNumberPassed *big.Int) error {
...
 }
 
+//Check if the block number ( date on which the token was issued is not more than 1 month)
+func CheckAllowedBlockDifferenceForToken(blockNumberPassed *big.Int) error {
+	latestBlockNumber, err := CurrentBlock()
+	if err != nil {
+		return err
+	}
+	differenceInBlockNumber := blockNumberPassed.Sub(blockNumberPassed, latestBlockNumber)
+	if differenceInBlockNumber.Abs(differenceInBlockNumber).Uint64() > AllowedBlockDifferenceForAuthorizedToken {
```
Comment (vsbogd) [inline authutils/auth_service.go]: Why don't use `big.Int.Cmp` instead?

File: authutils/auth_service.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ import (
 
 const (
 	AllowedBlockChainDifference = 5
+	//1 month is approximately 172800 blocks
+	AllowedBlockDifferenceForAuthorizedToken = 172800
```
Comment (vsbogd) [inline authutils/auth_service.go]: We may pass it with token instead of hardcoding it.

File: authutils/auth_service_test.go [hunk 3]
```diff
@@ -22,3 +22,16 @@ func TestCompareWithLatestBlockNumber(t *testing.T) {
 	assert.Equal(t, nil, err)
 
 }
+
+func TestCheckAllowedBlockDifferenceForToken(t *testing.T) {
+	config.Vip().Set(config.BlockChainNetworkSelected, "ropsten")
+	config.Validate()
+	currentBlockNum, _ := CurrentBlock()
+	err := CheckAllowedBlockDifferenceForToken(currentBlockNum.Add(currentBlockNum, big.NewInt(2000000)))
```
Comment (vsbogd) [inline authutils/auth_service_test.go]: As far as I understand the logic `CheckAllowedBlockDifferenceForToken` gets a date when token was issued in blocks. Then this date should be less then current date. But in test the date which bigger then current date is passed.
```suggestion
	err := CheckAllowedBlockDifferenceForToken(currentBlockNum.Sub(currentBlockNum, big.NewInt(2000000)))
```

File: escrow/free_call_api.go [hunk 4]
```diff
@@ -17,6 +17,13 @@ type FreeCallPayment struct {
 	CurrentBlockNumber *big.Int
 	// Signature passed
 	Signature []byte
+	//Group ID
+	GroupId string
+	//Auth Token Passed
+	AuthToken []byte
+	//Date on when the token was issued in block number
+	AuthTokenIssueDateBlockNumber *big.Int
```
Comment (vsbogd) [inline escrow/free_call_api.go]: BTW, why not pass expiration date instead? 
1. It doesn't require some expiration date hardcode on the daemon side. 
2. It makes it clear for client when token is expected to be expired.

File: authutils/auth_service_test.go (hunk 3 above)
Comment (anandrgitnirman) [reply authutils/auth_service_test.go]: now its Just Cmp

File: authutils/auth_service.go (hunk 2 above)
Comment (anandrgitnirman) [reply authutils/auth_service.go]: got away with this now , we just check if current block < tokens's Expiry block

File: authutils/auth_service.go (hunk 1 above)
Comment (anandrgitnirman) [reply authutils/auth_service.go]: done

File: authutils/auth_service.go [hunk 5]
```diff
@@ -83,6 +83,19 @@ func CompareWithLatestBlockNumber(blockNumberPassed *big.Int) error {
 	return nil
 }
 
+//Check if the block number ( date on which the token was issued is not more than 1 month)
+func CheckIfTokenHasExpired(blockNumberPassed *big.Int) error {
```
Comment (vsbogd) [inline authutils/auth_service.go]: I would rename `blockNumberPassed` to `expirationBlock`

File: escrow/free_call_api.go [hunk 6]
```diff
@@ -17,6 +17,13 @@ type FreeCallPayment struct {
 	CurrentBlockNumber *big.Int
 	// Signature passed
 	Signature []byte
+	//Group ID
+	GroupId string
+	//Auth Token Passed
+	AuthToken []byte
+	//Date on when the token was issued in block number
```
Comment (vsbogd) [inline escrow/free_call_api.go]:
```suggestion
	//Token expiration date in blocks
```

File: escrow/free_call_payment_handler.go [hunk 7]
```diff
@@ -71,17 +71,29 @@ func (h *freeCallPaymentHandler) getPaymentFromContext(context *handler.GrpcStre
 		return
 	}
 
+	authTokenIssueDate, err := handler.GetBigInt(context.MD, handler.FreeCallAuthTokenBlockNumberHeader)
```
Comment (vsbogd) [inline escrow/free_call_payment_handler.go]: Better to rename `authTokenIssueDate` to `authTokenExpiryDate`
And `FreeCallAuthTokenBlockNumberHeader` to `FreeCallAuthTokenExpiryBlockNumberHeader`

File: escrow/state_service.proto [hunk 8]
```diff
@@ -51,3 +51,31 @@ message ChannelStateReply {
...
+    rpc GetFreeCallsAvailable(FreeCallStateRequest) returns (FreeCallStateReply) {}
+}
+
+message FreeCallStateRequest {
+    //Has the user email id
+    string user_id = 1;
+    //signer-token = (user@mail, user-public-key, token_issue_date), this is generated my Market place Dapp
+    //to leverage free calls from SDK/ snet-cli, you will need this signer-token to be downloaded from Dapp
+    bytes token_for_free_call = 2;
+    //Token issue date in Block number
```
Comment (vsbogd) [inline escrow/state_service.proto]:
```suggestion
    //Token expiration date in Block number
```

File: escrow/validation_test.go [hunk 9]
```diff
@@ -37,12 +35,22 @@ func SignFreeTestPayment(payment *FreeCallPayment, privateKey *ecdsa.PrivateKey)
...
 		[]byte(config.GetString(config.ServiceId)),
+		[]byte(payment.GroupId),
 		bigIntToBytes(payment.CurrentBlockNumber),
+		payment.AuthToken,
 	}, nil)
 
 	payment.Signature = getSignature(message, privateKey)
 }
 
+func GenerateFreeCallTokenPayment(payment *FreeCallPayment, privateKey *ecdsa.PrivateKey,userAddress common.Address) {
```
Comment (vsbogd) [inline escrow/validation_test.go]: Just to confirm am I right that AuthToken includes only:
- user id
- user address
- token expiry block
?
It doesn't include orgId, serviceId and groupId?

File: resources/blockchain_network_config.json [hunk 10]
```diff
@@ -9,7 +9,7 @@
     "network_id":"42"
   },
   "ropsten":{
-    "ethereum_json_rpc_endpoint":"https://ropsten.infura.io",
+    "ethereum_json_rpc_endpoint":"https://ropsten.infura.io/v3/09027f4a13e841d48dbfefc67e7685d5",
```
Comment (vsbogd) [inline resources/blockchain_network_config.json]: Why not common SingularityNet project-id?

File: handler/interceptors.go [hunk 11]
```diff
@@ -50,6 +50,10 @@ const (
 	//Will be used to check if the Signature is still valid
 	CurrentBlockNumberHeader = "snet-current-block-number"
 
+	//Place holder to set the free call Auth Token issued
+	FreeCallAuthTokenHeader = "snet-free-call-auth-token-bin"
+    //Block number on when the Token was issued , to track the expiry of the token , which is ~ 1 Month
+    FreeCallAuthTokenBlockNumberHeader =  "snet-free-call-token-issue-block"
```
Comment (vsbogd) [inline handler/interceptors.go]:
```suggestion
    FreeCallAuthTokenBlockNumberHeader =  "snet-free-call-token-expiry-block"
```

File: escrow/validation_test.go (hunk 9 above)
Comment (anandrgitnirman) [reply escrow/validation_test.go]: this is correct, but when you make a call to state_service / actuall call to Daemon , you need to sign using orgid, serviceid and groupid , ( the other signture) . .. do you recommend adding this on token too?

File: resources/blockchain_network_config.json (hunk 10 above)
Comment (anandrgitnirman) [reply resources/blockchain_network_config.json]: kept this for testing alone ..

File: authutils/auth_service.go (hunk 5 above)
Comment (anandrgitnirman) [reply authutils/auth_service.go]: Done

Comment (vsbogd) [reply authutils/auth_service.go]: Excuse me but I don't see the change? May be you didn't push it?

File: escrow/validation_test.go (hunk 9 above)
Comment (vsbogd) [reply escrow/validation_test.go]: > do you recommend adding this on token too?

Not at all. It depends on requirements should user get new token for each service or can use one for all of them. My opinion the last will be simpler for user and for implementation so lets leave it as is.

File: authutils/auth_service.go (hunk 5 above)
Comment (anandrgitnirman) [reply authutils/auth_service.go]: Apologies, I missed changing the name of the parameter

**Pull request merged**